app/src/main/java/org/mozilla/focus/broadcastreceiver/DownloadBroadcastReceiver.java

```diff
@@ -0,0 +1,99 @@
+            DownloadManager downloadManager = referenceIds.get(downloadReference);
+            final DownloadManager.Query query = new DownloadManager.Query();
+            query.setFilterById(downloadReference);
+            try (Cursor cursor = downloadManager.query(query)) {
+                if (cursor.moveToFirst()) {
+                    int columnIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
+                    if (DownloadManager.STATUS_SUCCESSFUL == cursor.getInt(columnIndex)) {
+                        String uriString = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
+
+                        if (uriString.startsWith("file://")) {
+                            uriString = uriString.substring(7);
+                        }
+                        final String localUri = uriString;
+
+                        final String fileExtension = MimeTypeMap.getFileExtensionFromUrl(localUri);
+                        final String fileName = URLUtil.guessFileName(localUri, null, null);
+                        final String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension);
+
+                        final File file = new File(localUri);
+                        final Uri uriFile = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".fileprovider", file);
+
+                        final Intent openFileIntent = new Intent(Intent.ACTION_VIEW);
+                        openFileIntent.setDataAndType(uriFile, mimeType);
+                        openFileIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+                        final PackageManager packageManager = context.getPackageManager();
+                        final List<ResolveInfo> resolveInfoList = packageManager.queryIntentActivities(openFileIntent, 0);
+                        final boolean activitiesFoundForIntent = (resolveInfoList.size() > 0);
+
+                        final Snackbar snackbar = Snackbar
+                                .make(browserContainer, String.format(context.getString(R.string.download_snackbar_finished), fileName), Snackbar.LENGTH_LONG)
+                                .setAction(context.getString(R.string.download_snackbar_open), new View.OnClickListener() {
+                                    @Override
+                                    public void onClick(View view) {
+                                        if (activitiesFoundForIntent && mimeType != null) {
+                                            context.startActivity(openFileIntent);
+                                        }
+                                    }
+                                });
+                        if (activitiesFoundForIntent) {
+                            snackbar.setActionTextColor(ContextCompat.getColor(context, R.color.download_dialog_text));
+                        } else {
+                            snackbar.setAction("", null);
```

Comment:
nit: instead of setting the action to null here if there are no activities found, we should just never add the action in the first place